add-on/src/lib/redirect-handler/blockOrObserve.ts

```diff
@@ -0,0 +1,124 @@
+import browser from 'webextension-polyfill'
+import debug from 'debug'
+import { CompanionState } from '../../types/companion.js'
+
+const log = debug('ipfs-companion:redirect-handler:blockOrObserve')
+log.error = debug('ipfs-companion:redirect-handler:blockOrObserve:error')
+
+const savedRegexFilters = new Map<string, string>()
+
+interface redirectHandlerInput {
+  originUrl: string
+  redirectUrl: string
+}
+
+/**
+ * Construct a regex filter and substitution for a redirect.
+ *
+ * @param originUrl
+ * @param redirectUrl
+ * @returns
+ */
+function constructRegexFilter ({ originUrl, redirectUrl }: redirectHandlerInput): {
+  regexSubstitution: string
+  regexFilter: string
+} {
+  // these characters are allowed in the URL, but not in the regex.
+  // eslint-disable-next-line no-useless-escape
+  const ALLOWED_CHARS_URL_REGEX = /([:\/\?#\[\]@!$&'\(\ )\*\+,;=-_\.~])/g
+  // We can traverse the URL from the end, and find the first character that is different.
+  let commonIdx = 1
+  while (commonIdx < Math.min(originUrl.length, redirectUrl.length)) {
+    if (originUrl[originUrl.length - commonIdx] !== redirectUrl[redirectUrl.length - commonIdx]) {
+      break
+    }
+    commonIdx += 1
+  }
+
+  // We can now construct the regex filter and substitution.
+  const regexSubstitution = redirectUrl.slice(0, redirectUrl.length - commonIdx + 1) + '\\1'
+  // We need to escape the characters that are allowed in the URL, but not in the regex.
+  const regexFilterFirst = `${originUrl.slice(0, originUrl.length - commonIdx + 1).replace(ALLOWED_CHARS_URL_REGEX, '\\$1')}`
+  // We need to match the rest of the URL, so we can use a wildcard.
+  const regexFilter = `^${regexFilterFirst}(.*)$`
+
+  return { regexSubstitution, regexFilter }
+}
+
+// We need to check if the browser supports the declarativeNetRequest API.
+export const supportsBlock = !(browser.declarativeNetRequest?.MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES === 5000)
```

Comment:
do we need to check that the value is hardcoded to 5000? can we check that the value is defined to something instead? it should be more future proof

Reply:
My original comment here still applies. If chrome updates their value from 5000, this will fail. Do we want to have to keep an eye on their value, or can we make this check more robust?